src/preact/amp-base-element.js

```diff
@@ -0,0 +1,122 @@
+    return true;
+  }
+
+  // todo(kvchari): confirm this can be safely moved
+  /** @override */
+  mutatedAttributesCallback() {
+    // eslint-disable-next-line local/restrict-this-access
+    if (this.container_) {
+      // eslint-disable-next-line local/restrict-this-access
+      this.scheduleRender_();
+    }
+  }
+
+  // todo(kvchari): confirm this can be safely moved
+  /** @override */
+  attemptChangeHeight(newHeight) {
+    return super.attemptChangeHeight(newHeight).catch((e) => {
+      // It's okay to disable this lint rule since we check that the restricted
+      // method exists.
+      // eslint-disable-next-line local/restrict-this-access
+      if (this.getOverflowElement && !this.getOverflowElement()) {
+        console./* OK */ warn(
+          '[overflow] element not found. Provide one to enable resizing to full contents.',
+          this.element
+        );
+      }
+      throw e;
+    });
+  }
+
+  /** @override */
+  getDefaultProps() {
+    return dict({
+      // eslint-disable-next-line local/dict-string-keys
+      ...super.getDefaultProps(),
+      'onLoading': () => {
+        this.handleOnLoading();
+      },
+      'onLoad': () => {
+        this.handleOnLoad();
+      },
+      'onError': () => {
+        this.handleOnError();
+      },
+    });
+  }
+
+  /**
+   * Default handler for onLoad event
+   * Displays loader. Override to customize.
+   * @protected
+   */
+  handleOnLoad() {
+    this.toggleLoading?.(false);
+    this.toggleFallback?.(false);
+    this.togglePlaceholder?.(false);
+  }
+
+  /**
+   * Default handler for onLoading event
+   * Reveals loader. Override to customize.
+   * @protected
+   */
+  handleOnLoading() {
+    this.toggleLoading?.(true);
+  }
+
+  /**
+   * Default handler for onError event
+   * Displays Fallback / Placeholder. Override to customize.
+   * @protected
+   */
+  handleOnError() {
+    this.toggleLoading?.(false);
+    // If the content fails to load and there's a fallback element, display the fallback.
+    // Otherwise, continue displaying the placeholder.
+    if (this.getFallback?.()) {
+      this.toggleFallback?.(true);
+      this.togglePlaceholder?.(false);
+    } else {
+      this.togglePlaceholder?.(true);
+    }
+  }
+
+  /**
+   * Register an action for AMP documents to execute an API handler.
+   *
+   * This has no effect on Bento documents, since they lack an Actions system.
+   * Instead, they should use `(await element.getApi()).action()`
+   * @param {string} alias
+   * @param {function(!API_TYPE, !../service/action-impl.ActionInvocation)} handler
+   * @param {../action-constants.ActionTrust_Enum} minTrust
+   * @protected
+   */
+  registerApiAction(alias, handler, minTrust = ActionTrust_Enum.DEFAULT) {
+    this.registerAction?.(
+      alias,
+      (invocation) => handler(this.api(), invocation),
+      minTrust
+    );
+  }
+}
+
+/**
+ *
+ * @param {*} X
+ * @param {*} Y
+ * @return {*}
+ */
+export function setSuperClass(X, Y) {
```

Comment:
Drive-by bikeshed (don't spend too much time here): I know @alanorozco and I disagree. But wanted to note here that I'm not a fan of the `return X` line. The code ends up looking cleaner, but also gives the appearance of purity.

Reply:
Totally get what you mean - returning X does give the illusion that a new class is created, when in reality X's prototype is being modified (which will affect all instances of X whether they realize or not)

I could change

```
class AmpFoo extends setSuperClass(BaseElement, AmpPreactBaseElement) {...}
```

to 

```
setSuperClass(BaseElement, AmpPreactBaseElement)
class AmpFoo extends BaseElement {...}
```